[charlie] 카페 1,2단계 구현

kimm-charlie opened the pull request:
안녕하세요 Cooper, charlie입니다. 



Spring을 직접 다뤄본적이 없어 김영한님 강의를 보면서 미션을 진행했습니다. 

아직 에러처리 부분과 test코드를 작성하지 못했습니다. 양해부탁드립니다.

---

### 기능 목록

- 회원 가입 및 회원 목록 조회
  - 회원 가입 기능 
  - 회원 목록 조회 기능
  - 회원 프로필 조회 기능

- 글 쓰기 
  - 글쓰기 기능 
  - 글 목록 조회 기능
  - 게시글 상세보기 기능

- 회원정보 수정
  - 회원정보 수정 화면
  - 회원정보 수정 기능

### URL 

| 기능                  | URL                | Templates                   | HTTP 메서드 |
| --------------------- | ------------------ | --------------------------- | ----------- |
| 회원 가입 기능        | /user/join         | templates/user/form.html    | POST        |
| 회원 목록 조회 기능   | /users             | templates/user/list.html    | GET         |
| 회원 프로필 조회 기능 | /users/{id}        | templates/user/profile.html | GET         |
| 글쓰기 기능           | /post/submit       | templates/post/form.html    | POST        |
| 게시글 목록 조회 기능 | /                  | templates/index.html        | GET         |
| 게시글 상세보기 기능  | /articles/{id}     | templates/post/show         | GET         |
| 회원정보 수정 화면    | /users/{id}/form   | templates/user/updateForm   | GET         |
| 회원정보 수정 기능    | /users/{id}/update |                             | PUT         |

### 진행 흐름

![스크린샷 2023-03-31 오후 2.13.11](https://raw.githubusercontent.com/CDBchan/Typora-img/main/img/%E1%84%89%E1%85%B3%E1%84%8F%E1%85%B3%E1%84%85%E1%85%B5%E1%86%AB%E1%84%89%E1%85%A3%E1%86%BA%202023-03-31%20%E1%84%8B%E1%85%A9%E1%84%92%E1%85%AE%202.13.11.png)

출처: [DTO](https://e-una.tistory.com/72) 

클라이언트로부터 온 request를 DTO로 매핑후 원하는 동작을 하도록 했습니다. 



### 고민과 해결

1. DTO를 request 정보를 매핑하는 용도로만 사용하고 DTO로 매핑된 정보를  Controller에서 Service 계층으로 보내 원하는 처리를 하도록 했는데, 이렇게 하다보니 id 값을 entity의 생성자를 통해서 설정해줄수가 없었습니다.(id값 설정은 각 entity의 repository에서 설정을 해줬습니다.) 

   이때 DTO를 repository영역까지 가지고 가면 생성자를 통해 id값을 설정해줄수 있는데 생성자를 통해 id값을 설정해주는 방법과 위방법중 어떤게 더 좋은 방법인지 모르겠습니다. 일단은 위 그림을 준수하기위해 앞서 말한 방식(id값을 repository에서 설정)을 사용했습니다.    

   > ❗️헉.. id를 회원가입할때 입력을 해주는것 같습니다. 그렇다면 위와같은 문제가 발생하지 않을텐데.. 일단 지금방식으로 진행을하고 추후 수정을 하겠습니다. 

2. Service 계층에서 로직을 직접 처리하지않고 entity를 저장 또는 수정 하는 기능은 repository에 구현했습니다. 따라서 Service는 기능을 수행하기위해 Repository의 특정 메서드를 호출하는 방식으로 진행을 했고, 추후 필요한 예외처리에 대한 로직은 각 Service class에서 할 생각입니다. 이게 올바른 방법인지 궁금합니다.

   

## 실행결과

<details markdown="1">
<summary>회원가입</summary>


회원가입에 필요한 이메일,닉네임,비밀번호를 입력받는다.
![회원가입](https://raw.githubusercontent.com/CDBchan/Typora-img/main/img/%E1%84%92%E1%85%AC%E1%84%8B%E1%85%AF%E1%86%AB%E1%84%80%E1%85%A1%E1%84%8B%E1%85%B5%E1%86%B8.png)

</details>

<details markdown="1">
<summary>회원목록 조회</summary>


회원가입 완료후 가입을한 모든 회원을 보여준다.(지금 예시는1명이지만 회원가입을 계속해서하면 회원가입한 멤버가 모두 표시됩니다!)
![회원 목록 조회](https://raw.githubusercontent.com/CDBchan/Typora-img/main/img/%E1%84%92%E1%85%AC%E1%84%8B%E1%85%AF%E1%86%AB%20%E1%84%86%E1%85%A9%E1%86%A8%E1%84%85%E1%85%A9%E1%86%A8%20%E1%84%8C%E1%85%A9%E1%84%92%E1%85%AC.png)

</details>

<details markdown="1">
<summary>특정 회원 프로필 조회</summary>


회원목록에서 특정회원의 닉네임을 누르면 해당 회원의 프로필로 갈수있다.
![특정 회원 프로필 조회](https://raw.githubusercontent.com/CDBchan/Typora-img/main/img/%E1%84%90%E1%85%B3%E1%86%A8%E1%84%8C%E1%85%A5%E1%86%BC%20%E1%84%92%E1%85%AC%E1%84%8B%E1%85%AF%E1%86%AB%20%E1%84%91%E1%85%B3%E1%84%85%E1%85%A9%E1%84%91%E1%85%B5%E1%86%AF%20%E1%84%8C%E1%85%A9%E1%84%92%E1%85%AC.png)

</details>

<details markdown="1">
<summary>글쓰기 기능</summary>


메인페이지에서 글쓰기 버튼을 통해 글을쓸수있는 form으로 이동할수있다.
![글쓰기 기능](https://raw.githubusercontent.com/CDBchan/Typora-img/main/img/%E1%84%80%E1%85%B3%E1%86%AF%E1%84%8A%E1%85%B3%E1%84%80%E1%85%B5%20%E1%84%80%E1%85%B5%E1%84%82%E1%85%B3%E1%86%BC.png)


</details>

<details markdown="1">
<summary>게시글 목록 조회 기능</summary>


글쓰기 기능이 완료된후 자동적으로 메인페이지로 가게되고 이때 메인페이지에 내가 작성한 글이 표시된다.
![게시글 목록 조회 기능](https://raw.githubusercontent.com/CDBchan/Typora-img/main/img/%E1%84%80%E1%85%A6%E1%84%89%E1%85%B5%E1%84%80%E1%85%B3%E1%86%AF%20%E1%84%86%E1%85%A9%E1%86%A8%E1%84%85%E1%85%A9%E1%86%A8%20%E1%84%8C%E1%85%A9%E1%84%92%E1%85%AC%20%E1%84%80%E1%85%B5%E1%84%82%E1%85%B3%E1%86%BC.png)


</details>

</details>

<details markdown="1">
<summary>게시글 상세보기 기능</summary>


메인페이지에서 글의 제목을 누르면 해당글의 상세 내용을 볼수있다.(지금은 제목과 내용만 업데이트 됩니다!)
![게시글 상세보기 기능2](https://raw.githubusercontent.com/CDBchan/Typora-img/main/img/%E1%84%80%E1%85%A6%E1%84%89%E1%85%B5%E1%84%80%E1%85%B3%E1%86%AF%20%E1%84%89%E1%85%A1%E1%86%BC%E1%84%89%E1%85%A6%E1%84%87%E1%85%A9%E1%84%80%E1%85%B5%20%E1%84%80%E1%85%B5%E1%84%82%E1%85%B3%E1%86%BC2.png)



</details>

<details markdown="1">
<summary>회원정보 수정 화면</summary>


특정 멤버의 프로필에 들어가 오른쪽 위 회원정보 수정 버튼을 통해 회원정보 수정 화면으로 이동할수 있다.
이때 기존비밀번호 입력란에 입력한 비밀번호와 서버 repository에 저장된 비밀번호가 다르다면 회원정보를 바꿀수 없다.
![회원정보 수정 화면](https://raw.githubusercontent.com/CDBchan/Typora-img/main/img/%E1%84%92%E1%85%AC%E1%84%8B%E1%85%AF%E1%86%AB%E1%84%8C%E1%85%A5%E1%86%BC%E1%84%87%E1%85%A9%20%E1%84%89%E1%85%AE%E1%84%8C%E1%85%A5%E1%86%BC%20%E1%84%92%E1%85%AA%E1%84%86%E1%85%A7%E1%86%AB.png)


</details>



</details>

<details markdown="1">
<summary>회원정보 수정후 화면</summary>

회원정보를 수정하면 자동적으로 멤버리스트 페이지로 이동하고 변경된 멤버의 닉네임과 이메일을 볼수있다.
![회원정보 수정후 화면](https://raw.githubusercontent.com/CDBchan/Typora-img/main/img/%E1%84%92%E1%85%AC%E1%84%8B%E1%85%AF%E1%86%AB%E1%84%8C%E1%85%A5%E1%86%BC%E1%84%87%E1%85%A9%20%E1%84%89%E1%85%AE%E1%84%8C%E1%85%A5%E1%86%BC%E1%84%92%E1%85%AE%20%E1%84%92%E1%85%AA%E1%84%86%E1%85%A7%E1%86%AB.png)


</details>

kimm-charlie committed:
docs: html,css 파일 변경


기존에 있는 html 대신 "K사 벡엔드 신입이 만든 마크업 파일" html로 바꿧다.

kimm-charlie committed:
build: build.gradle에 mustach 의존성 추가

kimm-charlie committed:
chore: 동적으로 생성할 html파일 templates 디렉토리로 이동

kimm-charlie committed:
docs: mustach 설정 추가

kimm-charlie committed:
docs: mustache 한글깨짐현상 수정

kimm-charlie committed:
feat: 회원가입 기능 구현

User, UserController, UserRepository 클래스를 만들어 회원가입 기능을 구현했다.

kimm-charlie committed:
feat: 회원 목록 조회 기능 구현

kimm-charlie committed:
feat: 회원 프로필 조회 기능 구현

list.html에 존재하는 닉네임 클릭시 특정 회원 프로필을 볼수있도록 구현했다.

kimm-charlie committed:
docs: profile.html 작성

profile.html파일이 없어서 간단하게 만들었다.
docs: profile.html에 nickName과 email 추가

동적으로 html에 nickName과 email이 추가될수 있도록 했다.

kimm-charlie committed:
refactor: html 중복요소 제거

fragments디렉토리를 만들어 html에서 중복되는 각 각 html에서 중복되는 header, navbar를 각각 하나의 html 파일로 만들었다. 

kimm-charlie committed:
feat: UserDto, Service 추가후 이전기능 구현

DTO를 사용해서 데이터를 이동시키는것이 entity를 사용하는것보다 안전하다 생각해 DTO를 추가했다.
비지니스로직은 Controller가 아닌 Service계층에서 수행되도록 Service 계층을 구현했다.

kimm-charlie committed:
refactor: html 내부 css경로 수정

kimm-charlie committed:
feat: MvcConfig 추가

MvcConfig를 추가한다. 이때 로직이 없는 매핑 메서드또한 MvcConfig에 추가후 Controller에서는 삭제했다.

kimm-charlie committed:
feat: 글쓰기 기능 구현

kimm-charlie committed:
feat: 글 목록 조회하기 기능 구현

kimm-charlie committed:
feat: 게시글 상세보기 기능 구현

kimm-charlie committed:
refactor: html 버튼 href 수정

kimm-charlie committed:
refactor: Dto 클래스의 id필드 제거 및 데이터저장 collection 변경

Dto에 id필드가 있을 필요가 없다. 따라서 필드에서 제거한다.
기존엔 List Collection을 사용해 Repository를 구현했는데 이것보다 id를 key 값으로 하는 Map을 사용한다.

kimm-charlie committed:
docs: 회원정보 수정을 위한 updateForm 작성

kimm-charlie committed:
refactor: UserDto  id필드 추가

user의 정보를 UserRepository에 있는 save메서드를 통해 update 하기위해서는 Dto에 id가 있어야 우리가 원하는형식으로 Map에 저장할수 있어 UserDto에 id를 추가한다.

kimm-charlie committed:
docs: Put메소드를 사용하기위한 설정 추가

kimm-charlie committed:
feat: profileEditDto추가

이전패스워드와 변경할 패스워드를 가진 DTO가 필요해 profileEditDto를 생성했다.

kimm-charlie committed:
feat: 회원정보 수정 기능 구현

kimm-charlie committed:
docs: 매핑에 따른 html 수정

kimm-charlie committed:
refactor: Mvc config에서 user/form 매핑 삭제 

MvcConfig에서 /user/form URL을 /user/join 으로 GetMapping하고 UserController 에서 /user/join으로 postMapping시 오류가 발생해 /user/form을 Getmapping 해주는 메서드는 UserController로 이동했습니다.

kimm-charlie committed:
refactor: DTO 패키지 변경

DTO 를 controller 패키지 내부로 이동했습니다.

kimm-charlie committed:
refactor: DTO, Controller 패키지로 이동

kimm-charlie committed:
refactor: DTO는 client와 controller 간 통신에서만 사용하도록 수정

UserJoinDTO를 추가하고 기존 UserDTO를 삭제했습니다.

pbg0205 commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,33 @@
+        return "user/form";
+    }
+
+    @PostMapping("user/join")
```
join 대신 `sign-up` 은 어떤가요??

pbg0205 commented on src/main/java/kr/codesqaud/cafe/controller/ArticleController.java:
```diff
@@ -0,0 +1,31 @@
+
+    private ArticleService articleService;
+
+    public ArticleController(ArticleService articleService) {
+        this.articleService = articleService;
+    }
+
```
생성자 주입👍👍 스프링에서도 생성자 방식을 권장하죠! 그렇다면 생성자 방식을 권장하는 이유는 무엇일까요?

pbg0205 commented on src/main/java/kr/codesqaud/cafe/controller/ArticleController.java:
```diff
@@ -0,0 +1,31 @@
+@Controller
+public class ArticleController {
+
+    private ArticleService articleService;
```
필드의 불변을 위해 final keyword 를 추가하면 좋을 것 같습니다

kimm-charlie replied:
join 보다 sign-up이 회원가입이라는 기능을 나타내기에 더 적합한것 같아 변경했습니다!

pbg0205 commented on src/main/java/kr/codesqaud/cafe/controller/ArticleController.java:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    @PostMapping("/post/submit")
+    public String postArticle(ArticleDTO articleDto) {
```
post 대신 article 로 맞추면 통일성을 유지할 수 있을 것 같습니다 👍

pbg0205 commented on src/main/java/kr/codesqaud/cafe/service/ArticleService.java:
```diff
@@ -0,0 +1,28 @@
+    private ArticleRepository articleRepository;
+
+    public ArticleService() {
+        this.articleRepository = new ArticleRepository();
```
ArticleRepository 도 빈으로 선언해서 생성자 주입으로 변경하면 좋을 것 같아요 😃

pbg0205 commented on src/main/java/kr/codesqaud/cafe/repository/ArticleRepository.java:
```diff
@@ -0,0 +1,36 @@
+        return articleRepository.values().stream()
+                .filter(article -> article.getId() == id)
+                .findFirst()
+                .orElse(null);
```
null 대신 예외를 발생시키고 이후에 예외를 추가하는건 하는건 어떤가요?
예외처리는 추후 작업하는걸로 하고요!

kimm-charlie replied:
ArticleNotFoundException, PasswordNotFoundException, UserNotFoundException을 추가했습니다. 아직 에러가 발생하더라도 페이지가 넘어가지 않는 작업은 못했습니다. js를 통해 추후 작업 하겠습니다

pbg0205 commented on src/main/java/kr/codesqaud/cafe/repository/ArticleRepository.java:
```diff
@@ -0,0 +1,36 @@
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class ArticleRepository {
```
Repository 도 빈으로 등록시키면 좋을 것 같아요!
(확인해보기 : 자바로 빈을 등록하는 방법은 두가지 방식이 있죠! 그렇다면 스프링부트는 어떻게 빈을 등록할까요? @SpringBootApplication 에 @ComponentScan 이 있어 기본적으로 클래스의 하위를 탐색해 빈을 등록하기 때문인데요. 빈 등록에 관련해서 학습하면 한번 확인만 해보시고 아직 학습하지 않으셨다면 이후에 확인해보셔도 됩니다ㅎㅎ)

kimm-charlie replied:
스프링부트에서 빈을 등록하는 방법에 대해 알아보겠습니다!

pbg0205 replied:
@SpringBootApplication 에 @ComponentScan 에 어노테이션이 있는 것만 확인해보시고 빈 선언 방식에 관해서만 학습하시면 될 것 같아요!

pbg0205 commented on src/main/java/kr/codesqaud/cafe/controller/MainController.java:
```diff
@@ -0,0 +1,21 @@
+
+@Controller
+public class MainController {
+    ArticleService articleService;
```
private 접근제한자 추가해주세요~ (+ final)

pbg0205 commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,58 @@
+        return "/user/updateForm";
+    }
+
+    @PutMapping("/users/{id}/update")
```
Put HTTP Method 에 수정이라는 내용이 포함되어 있어 update 는 제거해도 될 것 같습니다

kimm-charlie replied:
```java 
   @PutMapping("/users/{id}")
    public String updateUserData(@PathVariable String id, ProfileEditDTO profileEditDto){
        userService.updateUserByUserId(profileEditDto,id);
        return "redirect:/users";
    }

    @GetMapping("/users/{id}")
    public String showUserProfile(@PathVariable String id, Model model) {
        model.addAttribute("user", userService.getUserByUserId(id));
        return "user/profile";
    }
```
cooper 가 추천해주신대로 update는 제거했습니다. 근데 이렇게하면  `@PutMapping("/users/{id}")` 과 `@GetMapping("/users/{id}")`의 mapping url이 같은데 상관없을까요? 일단은 mapping URL이 같더라도 @PutMapping, @GetMapping 만으로 어떤 역활을 할지 설명되는거 같아 위와같이 코드를 짜긴 했습니다.

pbg0205 replied:
스프링에서는 URL이 같더라도 Http Method 가 다르면 다른 것으로 판단하여 매핑할 수 있습니다.
스프링에서는 `HandlerMapping` 이라는 컴포넌트를 통해서 동작할 컨트톨러를 찾고 실행할할 수 있는 `HandlerAdapter` 를 찾습니다.
실제 스프링 코드를 보면 `RequestMappingHandlerMapping` 이라는 HandlerMapping 에서 매핑할 정보를 생성하고 매핑해서 URL 이 같아도 다른 것으로 구분해서 매핑하고 있어요! 이 부분은 질문주셔서 저도 찾아본 부분이라 깊게 공부해보시기 보다는 전체적인 MVC 구조에 학습하시고 나서 찾아보시는 걸 추천드려요👍👍 (덕분에 저도 공부했습니다 😃)

```java
public class RequestMappingHandlerMapping extends RequestMappingInfoHandlerMapping implements MatchableHandlerMapping, EmbeddedValueResolverAware {
        // 매팽 정보 생성
	protected RequestMappingInfo createRequestMappingInfo(
			RequestMapping requestMapping, @Nullable RequestCondition<?> customCondition) {

		RequestMappingInfo.Builder builder = RequestMappingInfo
				.paths(resolveEmbeddedValuesInPatterns(requestMapping.path()))
				.methods(requestMapping.method()) // http method
				.params(requestMapping.params())
				.headers(requestMapping.headers())
				.consumes(requestMapping.consumes())
				.produces(requestMapping.produces())
				.mappingName(requestMapping.name());
		if (customCondition != null) {
			builder.customCondition(customCondition);
		}
		return builder.options(this.config).build();
	}
      
}
```

pbg0205 commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,58 @@
+        return "/user/form";
+    }
+
+    @PostMapping("/user/join")
```
Post 에는 생성이라는 행위에 관한 내용을 내포하고 있어서 join 이 없어도 괜찮을 것 같아요

pbg0205 commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,58 @@
+@Controller
+public class UserController {
+
+    UserService userService;
```
private 접근제한자가 없어요😱 (이왕이면 final 도 같이 추가도 좋을 것 같아요!)

pbg0205 commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,58 @@
+
+    UserService userService;
+
+    public UserController() {
+        this.userService = new UserService();
+    }
+
```
UserService 는 빈으로 등록해서 생성자 주입 방식으로 사용하시면 좋을 것 같아요~

pbg0205 commented on src/main/java/kr/codesqaud/cafe/controller/ArticleController.java:
```diff
@@ -0,0 +1,31 @@
+
+    @GetMapping("/articles/{id}")
+    public String showDetailArticle(@PathVariable int id, Model model) {
+        model.addAttribute("article", articleService.findArticleById(id));
```
응답용 객체는 DTO 를 권장하고 있는데요. 응답용 객체를 DTO 로 전달하는 이유는 무엇일까요?
아래 글 한번 참고해서 학습해보시면 좋을 것 같아요 😃

- https://tecoble.techcourse.co.kr/post/2020-08-31-dto-vs-entity/

kimm-charlie replied:
감사합니다! DTO로 전달하는 이유를 공부해보겠습니다!

pbg0205 commented on src/main/java/kr/codesqaud/cafe/domain/User.java:
```diff
@@ -0,0 +1,54 @@
+
+    public void setPassword(String password) {
+        this.password = password;
+    }
```
사용하지 않는 코드는 제거해주세요. 다른 부분도 한번 확인 부탁드려요~

pbg0205 commented on src/main/java/kr/codesqaud/cafe/repository/UserRepository.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    public void save(User user) {
+        user.setId(user.getId() == -1 ? sequence++ : user.getId());
```
Article 하고 아이디 할당하는 방식을 동일하게 가져가면 되지 않을까요??
혹시 의도한 부분이면 말씀해주세요!

CDBchan replied:
UserRepositoy의 save메서드를 통해 바뀐 User의 설정정보를 저장하라는 요구사항이 있어 위와같이 했습니다.  
Article과 같은 방식으로 save를 하게되면 설정정보를 바꾼 user의 id가 수정되 
save할 user 객체의 id가 기본값인 -1 인 경우에만 새로운 id를 할당받도록 했습니다.

kimm-charlie replied:
- user의 id를 client로부터 받아오도록 기능을 변경했습니다! 따라서 아래 로직은 삭제했습니다.
```java
user.setId(user.getId() == -1 ? sequence++ : user.getId());
```

pbg0205 commented on src/main/java/kr/codesqaud/cafe/MvcConfig.java:
```diff
@@ -0,0 +1,22 @@
+        registry.setOrder(Ordered.HIGHEST_PRECEDENCE);
+
+        //user
+//        registry.addViewController("/user/join").setViewName("/user/form");
```
사용하지 않는 주석은 제거해주세요

pbg0205 commented on src/main/java/kr/codesqaud/cafe/MvcConfig.java:
```diff
@@ -0,0 +1,22 @@
+import org.springframework.web.servlet.config.annotation.WebMvcConfigurerAdapter;
+
+@Configuration
+public class MvcConfig extends WebMvcConfigurerAdapter {
```
spring 5.0 부터 `WebMvcConfigurer` 에 default method 로 addViewControllers 가 생겨서 `WebMvcConfigureAdapter` 대신 사용하시면 좋을 것 같습니다 😁

pbg0205 reviewed (changes requested):
안녕하세요 찰리~ 5주간 함께하게 된 리뷰어 쿠퍼 입니다 😃
같이 티키타카하면서 재밌게 미션 진행했으면 좋겠어요!
우선 아래에 질문이나 전체적인 리뷰 피드백에 관해 답변 드릴게요!

### 1. README.md 정리

`README.md` 에는 기본적인 프로젝트 기능 요구사항에 관해 설명 추가하는 용도로 사용됩니다. 찰리가 보내준 PR 보니까 
잘 정리하셔서 README.md 에 **기능 목록**, **URL** 복붙해서 추가하면 좋을 것 같아요

### 2. Layered Architecture 고민

> Service 계층에서 로직을 직접 처리하지않고 entity를 저장 또는 수정 하는 기능은 repository에 구현했습니다. 따라서 Service는 기능을 수행하기위해 Repository의 특정 메서드를 호출하는 방식으로 진행을 했고, 추후 필요한 예외처리에 대한 로직은 각 Service class에서 할 생각입니다. 이게 올바른 방법인지 궁금합니다.

 일반적인 Layered Architecture 의 경우에는 `Controller - Service -  Repository` 로 계층을 구분하는 방식이죠~ Service Layer 는 비즈니스 로직을 처리하는 부분이고 Repository 는 DB 에 저장, 조회를 하는 역할이기 떄문에 Service 에서 Repository 에게 위임하는 방식은 객체지향적인 관점에서 좋은 접근이라고 생각합니다. 잘 해주셨어요 👍👍

---

전체적으로 기본 틀은 잘 갖춰주신 것 같습니다😄 테스트나 예외처리에 관한 부분은 이후 미션들에서 천천히 작업해봐요~
피드백 드린 부분에 대해서 재정비만 해주시고 review request 보내주세요~

kimm-charlie committed:
refactor: join -> signUp

join보다는 signUp이 회원가입에 대한 의미를 나타내는데 적합하다 생각해 수정했습니다.

kimm-charlie committed:
refactor: Controller 필드 final로 선언

kimm-charlie committed:
refactor: 통일성을 위해 /post/submit -> /article/post 로  mapping URL 변경

kimm-charlie committed:
refactor: repository를 Spring bean으로  등록

kimm-charlie committed:
feat: UserNotFouddException, ArticleNotFoundException 추가

기존엔 찾아야 하는 user나 article이 없다면 null을 반환하도록했는데 null 대신 exception을 반환하도록 했다.

kimm-charlie committed:
refactor: /users/{id}/update -> /users/{id}

update에  대한 의미는 PutMapping에 들어나 있다. 따라서 mapping URL에서는 update에 대한 내용을 지운다.

kimm-charlie committed:
refactor: UserService bean등록

kimm-charlie committed:
refactor: DTO 또는 domain에서 사용하지 않는 setter getter 삭제

kimm-charlie committed:
refactor: WebConfigurerAdapter -> WebMvcConfigurer

registry.setOrder도 필요어 삭제했습니다.

kimm-charlie committed:
refactor: Repository interface 추가

지금은 Memory에 User 또는 Article을 저장하지만 추후 DB등을 사용할수 있다. 따라서 Repositoy interface를 만들어 코드를 더 확장성 있게 수정했다.

kimm-charlie committed:
feat: 회원가입시 userId입력 기능 구현

kimm-charlie committed:
refactor: LinkedHashMap을 사용해 저장순서를 보장

kimm-charlie committed:
test: ArticleService 에 대한 테스트 코드 작성

static 변수인 sequence를 초기화 하기위해 static 메서드인 resetSequenceForTest 메서드를 추가했습니다.

kimm-charlie committed:
feat: IDNotFoundException 추가

kimm-charlie committed:
test: UserService에 대한 테스트 코드 작성

kimm-charlie committed:
docs:README 업데이트

kimm-charlie committed:
refactor: IdNotFoundException -> PasswordNotFoundException

kimm-charlie commented:
### refactoring

- cooper가 추천해주신 방법으로 코드를 수정했습니다. 수정할때 의문점이 드는 부분이 몇가지 있었는데 질문드립니다.
1. MvcConfig
MvcConfig를 사용할때 발생한 오류

```java
@GetMapping("/user/signUp")
public String showSignUpForm(){
    return "/user/form";
}

@PostMapping("/user/signUp")
public String signUp(UserSignUpDTO userSignUpDto) {
    userService.addUser(userSignUpDto.toUser());
    return "redirect:/users";
}
```

UserConfroller에 위와같이 매핑을 하는 메서드가 있습니다. 이때 제가 알기론 메서드를 호출하지 않는 Mapping(위에선 showSignUpForm)에 대해서는 controller에 존재하는 showSignUpForm을 삭제하고 아래와같이 MvcConfig에 매핑을 해도 잘 동작을 하는줄알았는데 `Resolved [org.springframework.web.HttpRequestMethodNotSupportedException: Request method 'GET' not supported]`오류가 발생해 질문드립니다. 

```java
@Configuration
public class MvcConfig implements WebMvcConfigurer {
    @Override
    public void addViewControllers(ViewControllerRegistry registry) {
        
        registry.addViewController("/user/signUp").setViewName("/user/form");
    }
}

```

제가 생각하기론 `/user/signUp` URL을 호출했을때 MvcConfig에선 `@GetMapping`과 같은 mapping할 방법을 명시해주지 않아서 위와같은 오류가 발생한거고, 따라서 같은 URL에 대해 mapping을 하는 메서드가 여러개 있을땐 MvcConfig를 사용하지 않고 controller에서 mapping을 해야한다 라고 생각하고 있습니다. 

2. testCode 작성시 DisplayName()에 어떠한 내용을 적어야 하는지 감이 잘 안잡히는것 같습니다. 제가 작성한 방식대로 하면 되는지 궁금합니다.

pbg0205 commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,58 @@
+
+    public UserController() {
+        this.userService = new UserService();
+    }
```
Service 의 경우 생성자 주입으로 변경해주세요~

kimm-charlie replied:
service와 controller 모두 생성자 주입 방식으로 바꿧습니다!

pbg0205 commented on src/test/java/kr/codesqaud/cafe/service/UserServiceTest.java:
```diff
@@ -0,0 +1,109 @@
+        userService.addUser(new User("charlie3","aaa@naver.com","password","testId"));
+
+        //when & then
+        Assertions.assertFalse(userService.getUserList().size() == 4);
```
assertj 의 `Assertions.assertThat()` 을 사용하는 것을 추천드립니다

pbg0205 commented on src/test/java/kr/codesqaud/cafe/service/UserServiceTest.java:
```diff
@@ -0,0 +1,109 @@
+    }
+
+    @Test
+    @DisplayName("저장되지 않는 user id를 검색했을때 예외가 발생한다.")
```
지금도 괜찮은 것 같고 `유저를 검색하지 실패할 경우 예외를 반환한다.` 정도로 id에 관한 세부 내용을 제거하는 방식으로 적으셔도 될 것 같아요

pbg0205 commented on src/test/java/kr/codesqaud/cafe/service/UserServiceTest.java:
```diff
@@ -0,0 +1,109 @@
+    @BeforeEach
+    void initUserService(){
+        userService = new UserService();
+    }
```
@BeforeEach 사용 👍

pbg0205 commented on src/main/java/kr/codesqaud/cafe/repository/impl/MemoryUserRepository.java:
```diff
@@ -0,0 +1,52 @@
+    public void updateUser(ProfileEditDTO profileEditDto,String id) {
+        String oriPassword = userRepository.get(id).getPassword();
+        if (!oriPassword.equals(profileEditDto.getOriPassword())) {
+            throw new PasswordNotFoundException("비밀번호가 일치하지 않습니다.");
```
PasswordNotFound 라는 의미가 조금 어색한 것 같아요. 입력한 비밀번호가 유효하지 않음을 의미하는 `InvalidPasswordExcpetion` 네이밍은 어떨까요?

kimm-charlie replied:
지금 보니 PasswordNotFoundException은 좀 어색한거 같네요.. InavlidPasswordException으로 수정했습니다.!

pbg0205 commented on src/main/java/kr/codesqaud/cafe/MvcConfig.java:
```diff
@@ -0,0 +1,15 @@
+@Configuration
+public class MvcConfig implements WebMvcConfigurer {
+    @Override
+    public void addViewControllers(ViewControllerRegistry registry) {
```
같은 URL 일 경우에는 HTTP Method 와 관계없이 우선 순위가 @RequestMapping 을 먼저 검색하기 때문에 검색하지 못합니다. 만약 addViewControllers 로 분리해서 사용하고자 하는 경우에는 Controller 의 URL 과 다르게 선언해야 할 것 같아요 

`ViewControllerRegistry.java`
![image](https://user-images.githubusercontent.com/48561660/229357673-82e79add-fcae-4f96-8205-7a8eedab4f54.png)

kimm-charlie replied:
오! 감사합니다! 지금보니 제가 충분히 찾을수 있는 내용이였을것 같은데  다음부터는 조금더 찾아보고 질문드리겠습니다. 감사합니다.

pbg0205 replied:
저도 몰랐던 부분이라 덕분에 찾아봤어요~ 같이 찾아볼 만한 부분은 같이 찾아보시죠😎

pbg0205 commented on src/main/java/kr/codesqaud/cafe/MvcConfig.java:
```diff
@@ -0,0 +1,15 @@
+    public void addViewControllers(ViewControllerRegistry registry) {
+        //user
+        registry.addViewController("/user/login").setViewName("/user/login");
+        registry.addViewController("/user/login_success").setViewName("/user/login_success");
```
추가적으로 URI 경로로는 _(언더바) 보다 -(하이픈) 을 권장하고 있습니다. REST API 에서 사용하는 방식이지만 일반적으로 사용하지 않는 것으로 알고 있어서 작업하실 때 참고하시면 좋을 것 같습니다. 아래 참고할 링크 같이 첨부해드릴게요!

- https://meetup.nhncloud.com/posts/92

pbg0205 commented on src/main/java/kr/codesqaud/cafe/repository/impl/MemoryArticleRepository.java:
```diff
@@ -0,0 +1,46 @@
+                .orElseThrow(() -> new ArticleNotFoundException("해당 글이 존재하지 않습니다."));
+    }
+
+    public static void resetSequenceForTest() {
+        sequence = 1;
+    }
+}
```
테스트를 위한 코드는 제거해주세요!

pbg0205 commented on src/main/java/kr/codesqaud/cafe/service/ArticleService.java:
```diff
@@ -0,0 +1,29 @@
+public class ArticleService {
+    private ArticleRepository ArticleRepository;
+
+    public ArticleService() {
+        this.ArticleRepository = new MemoryArticleRepository();
+    }
```
생성자 주입 방식으로 변경해주세요~

pbg0205 commented on src/main/java/kr/codesqaud/cafe/service/UserService.java:
```diff
@@ -0,0 +1,34 @@
+
+    private UserRepository memoryUserRepository;
+
+    public UserService() {
+        this.memoryUserRepository = new MemoryUserRepository();
+    }
```
생성자 주입으로 변경해주세요

pbg0205 commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,57 @@
+        return "/user/form";
+    }
+
+    @PostMapping("/user/signUp")
```
일관성을 위해 `users` 로 통일해주세요

pbg0205 commented on src/main/java/kr/codesqaud/cafe/controller/dto/ArticleDTO.java:
```diff
@@ -0,0 +1,33 @@
+        return title;
+    }
+
+    public void setTitle(String title) {
+        this.title = title;
+    }
```
사용하지 않는 setter 는 제거해주세요

pbg0205 commented on src/main/java/kr/codesqaud/cafe/controller/dto/ArticleDTO.java:
```diff
@@ -0,0 +1,33 @@
+        return content;
+    }
+
+    public void setContent(String content) {
+        this.content = content;
+    }
```
사용하지 않는 setter 는 제거해주세요

pbg0205 reviewed (approved):
수고 많으셨습니다 찰리~ 더 좋은 방식에 관해서 고민하시는 모습이 저도 많은 자극을 받으면서 리뷰할 수 있었어요 😃
질문주신 두가지 질문에 관해 코멘트로 남겨서 관련된 부분 참고하시면 좋을 것 같습니다. 추가 확인하면서 코멘트를 더 남겼는데 이 부분은 다음 스텝에서 반영해주시면 될 것 같습니다. 주말에도 작업하시느라 수고 많으셨고 다음 스텝에서 뵐게요 👋
